staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -159,7 +160,14 @@ type SharedInformer interface {
 	// be competing load and scheduling noise.
 	// It returns a registration handle for the handler that can be used to remove
 	// the handler again and an error if the handler cannot be added.
+	//
+	// TODO: logcheck:context // AddEventHandlerWithConfig should be used instead of AddEventHandlerWithResyncPeriod in code which supports contextual logging.
 	AddEventHandlerWithResyncPeriod(handler ResourceEventHandler, resyncPeriod time.Duration) (ResourceEventHandlerRegistration, error)
+	// AddEventHandler is a variant of AddEventHandler and AddEventHandlerWithResyncPeriod
+	// where additional optional parameters are passed in a struct. If no resync period
+	// is specified there, AddEventHandlerWithConfig behaves like AddEventHandler.
+	// The context is used for contextual logging.
+	AddEventHandlerWithConfig(ctx context.Context, handler ResourceEventHandler, config HandlerConfig) (ResourceEventHandlerRegistration, error)
```

Comment:
This `WithConfig` is used instead of adding a `WithContext` because it enables adding other optional parameters besides resync period without having to add another method.

```
Incompatible changes:
- ./tools/cache.SharedInformer.AddEventHandlerWithConfig: added
- ./tools/cache.SharedInformer.RunWithContext: added
- ./tools/cache.SharedInformer.SetWatchErrorHandlerWithContext: added 
```

Same as with `Controller`: probably shouldn't be implemented outside of Kubernetes?

Reply:
This is implemented outside of kube, especially for mocking. Controller-runtime for example has an instance.

Reply:
Is it acceptable to break that other code? Probably not.

The alternative is to leave `cache.SharedInformer` as it is and add a second `cache.SharedInformerCtx` with the new methods. The code which returns a `cache.SharedInformer` can return a `cache.SharedInformerCtx` (assignable to variables or fields of type `cache.SharedInformer`). The only API break for such producers is that the function signature changes.

The downside of this approach is that code accepting a `cache.SharedInformer` must continue to do so. It can optionally use the new method only after a checked type cast.

Reply:
<bikeshed>Isn't such a struct usually called `Options` in Kube, not `Config`?</bikeshed>

Reply:
https://golang.cafe/blog/golang-functional-options-pattern.html calls it "Config struct" and there are several "Config structs" in Kubernetes. I think both is used. Not important, so let's use `Options` and `AddEventHandlerWithContext`, in particular for consistency within `shared_informer.go` itself.

> The only API break for such producers is that the function signature changes.

And of course, controller-runtime does exactly the one thing that doesn't work with such a change (assign functions to variables) :cry:  That means we need additional NewShared* variants.

Most code doesn't call those directly, but rather indirectly through client-go (e.g. https://github.com/kubernetes/kubernetes/blob/5973accf487eef985393cde200e7a4f9d09bc2e0/staging/src/k8s.io/client-go/informers/core/v1/pod.go#L38).

Changing that leads us back to the problem that it might be an API break for some code. If we leave it like it is now, callers need to cast. They either must check that cast (typically dead code!) or do it unchecked (no compile time checking).

For now I am going with "unchecked type cast" in the one example that I have in this PR (`pkg/controller/resourceclaim/controller.go`), but I am not sure - see updated PR.

Perhaps breaking clients like controller-runtime which implement `cache.SharedInformer` is the lesser evil?

/cc @sbueringer @alvaroaleman

Reply:
I checked how badly the initial commit breaks controller-runtime. Only a test helper (`pkg/controller/controllertest/util.go`) is affected. After removing `pkg/controller/controllertest` and `pkg/cache/informertest`, `go build ./...` works.

I think that satisfy what consumers of controller-runtime want: they can take an old release of controller-runtime and the 1.32 client-go and the combination will work.

I suggest we go ahead with the initial commit and remove efa8146d2ae8378845a84c5f824f864871f4489e.

Reply:
> I think that satisfy what consumers of controller-runtime want

... unless those consumers import those packages:
https://grep.app/search?q=sigs.k8s.io/controller-runtime/pkg/controller/controllertest

:cry:

Reply:
Sounds okay from my side. We intentionally state in our docs that every CR minor version is only compatible with a specific client-go minor version:

> Every minor version of controller-runtime has been tested with a specific minor version of client-go. A controller-runtime minor version may be compatible with other client-go minor versions, but this is by chance and neither supported nor tested. In general, we create one minor version of controller-runtime for each minor version of client-go and other k8s.io/* dependencies.
> https://github.com/kubernetes-sigs/controller-runtime?tab=readme-ov-file#compatibility

Also happened in the past that we weren't entirely compatible with other client-go versions.

So seems okay to me to have a breaking change here.

Reply:
cc @alvaroaleman @vincepri

Reply:
Thanks @sbueringer for heads up. I've pushed a single commit with the original proposal (extending `SharedInformer`) and with the renames suggested by @sttts above.

Please take another look, it might be ready for merging now.

Reply:
Looks good to me from a CR perspective